libsql-wal/src/bottomless/storage/fs.rs

```diff
@@ -19,19 +30,60 @@ impl Storage for FsStorage {
         segment_data: impl crate::io::file::FileExt,
         segment_index: Vec<u8>,
     ) -> impl Future<Output = Result<()>> + Send {
-        todo!();
-        #[allow(unreachable_code)]
-        std::future::ready(Ok(()))
+        let key = format!(
+            "{}-{}-{}.segment",
+            meta.start_frame_no,
+            meta.end_frame_no,
+            meta.created_at.timestamp()
+        );
+
+        let path = self.prefix.join("segments").join(key);
+
+        let buf = Vec::with_capacity(segment_data.len().unwrap() as usize);
+
+        async move {
+            let (buf, res) = segment_data.read_exact_at_async(buf, 0).await;
+
+            tokio::fs::write(path, buf).await.unwrap();
+
+            Ok(())
+        }
     }
 
     async fn fetch_segment(
         &self,
         _config: &Self::Config,
         _namespace: NamespaceName,
-        _frame_no: u64,
-        _dest: impl AsyncWrite,
+        frame_no: u64,
+        dest: impl AsyncWrite,
     ) -> Result<()> {
-        todo!()
+        let dir = self.prefix.join("segments");
+
+        // TODO(lucio): optimization would be to cache this list, since we update the files in the
+        // store fn we can keep track without having to go to the OS each time.
+        let mut dirs = tokio::fs::read_dir(dir).await?;
+
+        while let Some(dir) = dirs.next_entry().await? {
+            let file = dir.file_name();
+            let key = file.to_str().unwrap().split(".").next().unwrap();
+            let mut comp = key.split("-");
+
+            let start_frame = comp.next().unwrap();
+            let end_frame = comp.next().unwrap();
+
+            let start_frame: u64 = start_frame.parse().unwrap();
+            let end_frame: u64 = end_frame.parse().unwrap();
+
+            if start_frame <= frame_no && end_frame >= frame_no {
+                let mut buf = tokio::fs::read(dir.path()).await.unwrap();
```

Comment:
Maybe a small assert around here to make sure the header matches the expected frame range?